src/github.com/getlantern/wfilter/wfilter.go

```diff
@@ -0,0 +1,66 @@
+// package wfilter provides facilities for adding filtering to io.Writer.
+package wfilter
+
+import (
+	"bytes"
+	"io"
+)
+
+type Prepend func(w io.Writer) (int, error)
+
+// lines creates a filtering writer that filters on a line by line basis using
+// the given filterFunc. FilterFunc is given the string for the current line
+// along with the original writer, and it can write whatever it wants. It should
+// return the bytes written and any error encountered.
+func LinePrepender(w io.Writer, prepend Prepend) io.Writer {
+	return &lp{w, prepend, true}
+}
+
+type lp struct {
+	io.Writer
+	prepend       Prepend
+	prependNeeded bool
+}
+
+func (w *lp) Write(buf []byte) (int, error) {
```

Comment:
I was trying very hard to understand this method but failed... So I write a shorter `do... while()` version, please take a look to check if any flaw exists.